### ISSUE TITLE
Evaluation & Metrics

### ISSUE THREAD

**Issue by jannikfranzen** [Kainmueller-Lab/Pathology-Foundation-Model-Benchmark#17]
*This is for internal use only; if you'd like to open an issue or request a new feature, please open a bug or enhancement issue*

**Relevant background**

Benchmarking different foundation models should be done on two main levels: 1) on the "instance level"; here we seek to train a simple linear layer + resizing to yield a pixelwise (but possibly blurry) semantic map and evaluate the cell type predictions by using regionprops on the predictions with the groundtruth instance masks, and 2) true pixel level where we seek to evaluate the semantic performance on the pixel level (which is required for properly evaluating the SegPath dataset).

On instance level: AUC, accuracy, f1 score, precision, recall 
On semantic level: AUC, accuracy, f1 score, precision, recall, iou
calculated per cell type individually

**Design overview**

The implementation of the metrics will be done in two files, one utils file containing the function of each individual metric and one file containing the metric computation pipeline as a class or function. After instanciation this pipeline should write the metrics into a dictionary and the instance wise predictions.

The instance level should be a priority for now. Here, we have an additional extra step to compute the semantic predictions on the instance level.

Metrics on instance level:

1. Accuracy
    - for all classes: correct/total
    - balanced accuracy
2. F1 score (and precision and recall)
    - for each class
    - average over all classes
        - micro average
        - macro average
3. Confusion matrix

additional metrics could be:

4. AUC-ROC: to evaluate the model's ability to distinguish between classes
5. AUC-PR: to evaluate the model's ability to predict positive classes
6. Mean Reciprocal Rank: measuring how high the true class is ranked in the predicted probabilities.

**Code mockup**

Metrics in the utility file:

```python
def accuracy(y_true, y_pred):
    """
    Args:
        y_true: 1D-array (n_cells), with integer values representing the true class
        y_pred: 1D-array (n_cells), with integer values representing the predicted class
    Returns:
        accuracy: float
    """
    pass

def precision(y_true, y_pred, class_id):
    """
    Args:
        y_true: 1D-array (n_cells), with integer values representing the true class
        y_pred: 1D-array (n_cells), with integer values representing the predicted class
    Returns:
        precision: float
    """
    pass

def recall(y_true, y_pred, class_id):
    """
    Args:
        y_true: 1D-array (n_cells), with integer values representing the true class
        y_pred: 1D-array (n_cells), with integer values representing the predicted class
    Returns:
        recall: float
    """

def f1_score(y_true, y_pred, class_id):
    """	
    Args:
        y_true: 1D-array (n_cells), with integer values representing the true class
        y_pred: 1D-array (n_cells), with integer values representing the predicted class
    Returns:
        f1_score: float
    """

def confusion_matrix(y_true, y_pred): # n_classes?
    """
    Args:
        y_true: 1D-array (n_cells), with integer values representing the true class
        y_pred: 1D-array (n_cells), with integer values representing the predicted class
    Returns:
        confusion_matrix: 2D-array (n_classes, n_classes)
    """
```

And for the class containing the pipeline:

```python
class Eval:
    def __init__(self, metrics, instance_level=True, pixel_level=False):
        self.metrics = metrics
        self.instance_level = instance_level
        self.pixel_level = pixel_level
        
    def compute_metrics(self, model, dataloader):
        """
        Args:
            model: torch.nn.Module
            dataloader: torch.utils.data.DataLoader
        Returns:
            metrics: dict
            instance_predictions: dict, empty if instance_level=False
        """
        instance_predictions = {}
        for  sample_dict in data_loader
            img = sample_dict["image"].float()
            semantic_mask = sample_dict["semantic_mask"]
            instance_mask = sample_dict.get("instance_mask", None)
            if self.instance_level:
                # calculate per cell prediction and GT with regionprops, add sample_dict['sample_name'] to the df and store dfs in a list to later concatenate them
            if self.pixel_level:
                pass  # not implemented errors
        
        if self.instance_level:
            metrics = self.compute_metrics_instance_level(inst_pred_class, inst_gt_class)
            
    
    def compute_metrics_instance_level(self, pred_prob, gt_class):
        """
        Args:
            pred_prob: 1D-array (n_cells, n_classes), with float values 
            representing the predicted class probabilities
            gt_class: 1D-array (n_cells), with integer values representing the true class
        Returns:
            metrics: dict
        """
        

    def compute_metrics_pixel_level(self, pred, gt):
        """
        Args:
            pred: np.array (H, W), with integer values representing the predicted class
            gt: np.array (H, W), with integer values representing the true class
        Returns:
            metrics: dict
        """
        pass
```

**Required inputs**

we need a unified structure for the dataloader __get_item__ outputs; rest see code mockup

**Output files**

Provide a description of the outputs for this project. If any plots will be generated, provide (simple) sketches demonstrating the plot type and axes labels. 

**Timeline**
Give a rough estimate for how long you think the project will take. In general, it's better to be too conservative rather than too optimistic. 
- [x] A couple days
- [ ] A week
- [ ] Multiple weeks. For large projects, make sure to agree on a plan that isn't just a single monster PR at the end. 

Estimated date when a fully implemented version will be ready for review:

Estimated date when the finalized project will be merged in:
